### PR TITLE
DEFAULT CONTAINER: Update docblock return type of has method

### PR DESCRIPTION
**Pull request by sjokkateer**
Signed-off-by: Remy Bos <27890746+sjokkateer@users.noreply.github.com>

|    Q          |   A
|-------------- | ------
| Documentation | yes
| Bugfix        | no
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | no

### Description
This PR updates the return type in the docblock for the `DefaultContainer::has()` method.
Currently it is said to return [mixed](https://github.com/laminas/laminas-di/blob/3.6.x/src/DefaultContainer.php#L60), while both the implemented `ContainerInterface::has()` is said to return a `bool`, as the method itself.

### CODE REVIEW

**Comment by sjokkateer** [comment]
I was also wondering if it is alright to update the return type of [DefaultContainer::get()](https://github.com/laminas/laminas-di/blob/3.6.x/src/DefaultContainer.php#L83) to the more specific `object` since it only seems to return objects?

Or should changing the return type of correct docblocks be considered breaking as well since static analysers might start complaining?

File: src/DefaultContainer.php [hunk 1]
```diff
@@ -57,7 +57,7 @@ public function setInstance(string $name, $service): self
      * @see ContainerInterface::has()
      *
      * @param string $name
-     * @return mixed
+     * @return bool
```
Comment (Ocramius) [inline src/DefaultContainer.php:60]: I actually wonder how this behaves with `psr/container:^2` - is omitting the return type allowed at all?

**Review by Ocramius — APPROVED** [review]
LGTM - simple and sweet improvement :-)

**Comment by Ocramius** [comment]
> since it only seems to return objects?

I remember the container being able to be populated with non-`object` types too?

**Comment by Ocramius** [comment]
Thanks @sjokkateer!

File: src/DefaultContainer.php (hunk 1 above)
Comment (sjokkateer) [reply src/DefaultContainer.php:60]: I tested the following, hopefully this is the setup you were wondering about:

`PHP 7.4.29 (cli)`
`psr/container": "^2.0"`

Class:

```php
<?php

declare(strict_types=1);

namespace App;

use Psr\Container\ContainerInterface;

/**
 * Undocumented class
 */
class Container implements ContainerInterface
{
    /**
     * Undocumented function
     */
    public function get(string $name)
    {
    }

    /**
     * Undocumented function
     */
    public function has(string $name)
    {
        return false;
    }
}
```

Script:
```php
<?php

use App\Container;

include __DIR__ . '/vendor/autoload.php';

$c = new Container();
```
Result:

`Fatal error: Declaration of App\Container::has(string $name) must be compatible with Psr\Container\ContainerInterface::has(string $id): bool`

Comment (Ocramius) [reply src/DefaultContainer.php:60]: Aye, guessed as much: a patch to upgrade to `psr/container:^2` will be messy and breaking BC too :-\

Comment (sjokkateer) [reply src/DefaultContainer.php:60]: That's unfortunate.

**Comment by sjokkateer** [comment]
> 



> > since it only seems to return objects?
> 
> I remember the container being able to be populated with non-`object` types too?

Hmm, I derived this from the constructor body and the docblocks within the class.

In the constructor only objects are added as services, an `InjectorInterface::class` and a `ContainerInterface::class`.
And in the [DefaultContainer::setInstance()](https://github.com/laminas/laminas-di/blob/3.6.x/src/DefaultContainer.php#L46) method, the docblock hints towards only accepting type `object` for the `$service` parameter, hence without considering how people in the meantime might have extended the `DefaultContainer::class` it seems to only assign and therefore return type `object`.

Perhaps this is unwanted, but I was wondering if :-P.

**Comment by sjokkateer** [comment]
> > since it only seems to return objects?
> 
> I remember the container being able to be populated with non-`object` types too?

Let me retry @Ocramius, in [DefaultContainer::__construct()](https://github.com/laminas/laminas-di/blob/3.6.x/src/DefaultContainer.php#L32) objects are registered as services.
Registering a new service through [DefaultContainer::setInstance()](https://github.com/laminas/laminas-di/blob/3.6.x/src/DefaultContainer.php#L46) is documented to only accept type `object` as param `$service`.
Further, in [DefaultContainer::get()](https://github.com/laminas/laminas-di/blob/3.6.x/src/DefaultContainer.php#L46) the forward call to the `InjectorInterface::class` implementation, according to the code of [InjectorInterface::create()](https://github.com/laminas/laminas-di/blob/3.6.x/src/InjectorInterface.php#L21) this method should return an `object`.

This makes it seem that all the branches return an `object` type, therefore the return type based on PHP 7.4's [covariance](https://www.php.net/manual/en/language.oop5.variance.php#language.oop5.variance)
> Covariance allows a child's method to return a more specific type than the return type of its parent's method.

could be made more specific to `object` instead of the current `mixed`, however, this might introduce issues with static analysis since users might have extended the `DefaultContainer::class` and rely on a `mixed` return type somehow?

**Comment by Ocramius** [comment]
The main problem with adding type declarations is that these classes are not `final` , and we would therefore break subclasses

**Comment by sjokkateer** [comment]
@Ocramius Thanks for your response but I am mainly wondering about return types in docblocks at the moment and would really appreciate if you could help take away my confusion.

Say in this case the docblock's return type would have been incorrect, would an update of the docblock's return type happen according to the same rules? By incorrect I would as an example assume that the author wanted to narrow down the return type in the docblock but forgot to do so.

**Comment by Ocramius** [comment]
Fine to add these to docblock! :+1:

**Comment by sjokkateer** [comment]
@Ocramius Thanks for the info!

**Pull request merged**